### DISCUSSION TITLE
Chained rate limiting policy

### DISCUSSION THREAD

**Discussion by SbiCA** [dotnet/aspnetcore#54051] [Q&A]
I was following the rate limiting docs https://learn.microsoft.com/en-us/aspnet/core/performance/rate-limit?view=aspnetcore-8.0#create-chained-limiters which allows to create a `GlobalLimiter` in a chained way. 

I'm missing this on policy level, or is there a way that I might have overlooked?

I would have expected something like this to work 

![image](https://github.com/dotnet/aspnetcore/assets/11393585/214c5caf-4410-444f-a7a8-2fedabce84ac)

My scenario is the following:

Existing application with lots of endpoints I don't want to enforce a global rate limit immediately since a reasonable default isn't known yet. Therefore I'd like to start lax and make the limits stricter once the load patter per policy is more clear.

Also note that in our case we have/want to distinguish certain workloads via policy because not all endpoints are treated equal e.g. authenticate/non-authenticated and post/get traffic

**Comment by SbiCA** [comment]
@sebastienros  or @danroth27  maybe? I'd like to understand if it's a user issue (highly likely) or a missing feature 😄

**Reply by juliuslouw** [reply]
I've been looking for this feature as well. We have a couple of endpoints in the solution that legitimately need some type of rate limit due to its cost. The other endpoints are fairly cheap and would be counterproductive to rate limit. 
I tried creating multiple policies - i.e., per 1s, per 1m, per 1h, per 1d - and apply all of them to one endpoint, but the `EnableRateLimiting` attribute constraints prevent this type of solution. 
A chained, partitioned rate limiting policy would be very useful.

**Reply by MikeNolan678** [reply]
Also (unsuccessfully) looking for a similar chained partitioned rate limiting feature. We impose higher limits on some internal endpoints currently, in a .Net Framework project using https://github.com/stefanprodan/WebApiThrottle. I'm moving this API to .Net 8.0, and would like to avoid using another NuGet package, but it looks like similar chaining is not possible unless it's applied to the GlobalLimiter.

**Reply by SbiCA** [reply]
seems like @JeremyLikness is would be able to tell more (at least that's what I understood from the last community standup 😄 )

**Reply by danroth27** [reply]
@BrennanConroy Can you take a look at this one?

**Reply by SbiCA** [reply]
@BrennanConroy did you give a look?

**Comment by BrennanConroy** [comment]
> I'd like to start lax and make the limits stricter once the load patter per policy is more clear

Maybe I'm misunderstanding, but how does a chained limiter solve this? It seems like you want to return a different rate limiter depending on dynamic data.

You can achieve this by changing the name of the partition and returning different settings in the new limiter:
```csharp
var dynamicData = 0;
if (dynamicData < 1)
{
    return RateLimitPartition.GetNoLimiter("noop");
}
else if (dynamicData < 5)
{
    return RateLimitPartition.GetConcurrencyLimiter("stricter", c =>
    {
        return new ConcurrencyLimiterOptions()
        {
            PermitLimit = 1000,
            QueueLimit = 100
        };
    });
}
else
{
    return RateLimitPartition.GetConcurrencyLimiter("strictest", c =>
    {
        return new ConcurrencyLimiterOptions()
        {
            PermitLimit = 100,
            QueueLimit = 10
        };
    });
}
```

> We have a couple of endpoints in the solution that legitimately need some type of rate limit due to its cost. The other endpoints are fairly cheap and would be counterproductive to rate limit.
I tried creating multiple policies - i.e., per 1s, per 1m, per 1h, per 1d - and apply all of them to one endpoint, but the EnableRateLimiting attribute constraints prevent this type of solution.

This is also weird and maybe a misunderstanding, why are you trying to apply all policies to all endpoints? If one endpoint is cheap, apply the cheap policy to it. If an endpoint is expensive, apply the expensive policy to it.
```csharp
app.MapGet("/one", () => { }).RequireRateLimiting("cheap");
app.MapGet("/two", () => { }).RequireRateLimiting("expensive");
```

**Reply by SbiCA** [reply]
@BrennanConroy  yes probably I wasn't clear enough. Let me try to give some more context.

I'd like to use the `ChainedRateLimiter` on endpoint level but somehow the policy build doesn't allow to use it (or I'm not getting how). 

Our scenario is more that the policy should
- Limit X requests per Min
- Limit Y requests per Hour
- Limit Z requests per Day

for a specific group of endpoints (hence we can't use the global rate limiter as shown in the docs)

``` c#

app.MapGet("/limited-one", () => { }).RequireRateLimiting("Policy-10perMin-20perHour-100perDay");
app.MapGet("/limited-two", () => { }).RequireRateLimiting("Policy-10perMin-20perHour-100perDay");
app.MapGet("/not-limited", () => { });

```

I hope that helps.

**Reply by BrennanConroy** [reply]
Sounds like you want https://github.com/dotnet/runtime/issues/98017 which should be "fairly easy" to implement yourself.

**Reply by SbiCA** [reply]
@BrennanConroy  I would say so yes.
But as a user there is one point that I don't get. 
Why I'm I allowed to use the following code when using `RateLimiterOptions` in `AddRateLimiter`

```c#
            // that works 
            limiterOptions.GlobalLimiter = PartitionedRateLimiter.CreateChained<HttpContext>( ... 
            
            // why can't I use the same here? 😕 
            // Option 1) ⬇️ maybe an overload that would allow to use `CreateChained` 
            limiterOptions.AddPolicy("some policy", context =>
            {
                var userId = GetAuthenticatedUserId(context);

                // Option 2) ⬇️  Allow to use `CreatedChained` here
                return RateLimitPartition.GetSlidingWindowLimiter(userId,
                    _ => new SlidingWindowRateLimiterOptions
                    {
                        PermitLimit = rateLimitsEventCreation.PermitLimit,
                        QueueLimit = rateLimitsEventCreation.QueueLimit,
                        Window = rateLimitsEventCreation.TimeSpan,
                        SegmentsPerWindow = rateLimitsEventCreation.SegmentsPerWindow
                    });
            });

```

That would be my natural way when seeing the current docs and I would try to use it. 
My problem with it being "fairly" easy is that:
1. the time it takes to figure that out would take a few hours (many types, what is tied to what builder, options, policies ..)
2. not to speak about coming up with a solution on my own (not having your value insights @BrennanConroy  but also that took weeks to get to)  

So IMHO that should become part of the default since it's already allowed on global level and it would improve the clarity of the feature

**Reply by SbiCA** [reply]
@danroth27  @BrennanConroy any chance this can make it into the 10 roadmap 🙏 ? It seems the foundation for it has been merged

**Comment by vanbukin** [comment]
@BrennanConroy 
Here’s an example from a real-world production scenario my team is working on:

**Context:**
We’re building an Identity Provider (IDP) using ASP.NET Core. One of our key requirements is implementing brute-force protection for login/password submissions.

**Goal:**
We need a unified chained rate-limiting policy with the following sliding window rules:
1) 5 segments, 5-minute window, 50 requests per username (extracted from the submitted form).
2) 24 segments, 24-hour window, 2000 requests per IP address.

If either limit is hit, the system should respond with HTTP 429 (Too Many Requests).
A chained policy would be an ideal fit for this use case.

**Reply by ApacheTech** [reply]
I have this exact scenario after a vulnerability report from penetration testing.

I have come up with a workaround that works for now, for IP and Email rate limiting for ForgotPassword attempts. It should work for you as well. But, I warn you, it's not overly pretty. Multiple policies per endpoint, or chained policies, would be much better.

https://github.com/dotnet/aspnetcore/issues/42691#issuecomment-3000791877
